pkg/sql/conn_executor.go

```diff
@@ -1390,6 +1380,22 @@ type connExecutor struct {
 		// createdSequences keeps track of sequences created in the current transaction.
 		// The map key is the sequence descpb.ID.
 		createdSequences map[descpb.ID]struct{}
+
+		// Fields that can be accessed outside connExecutors goroutine must
+		// be behind a mutex.
+		mu struct {
+			syncutil.RWMutex
+
+			// autoRetryReason records the error causing an auto-retryable error event if
+			// the current transaction is being automatically retried. This is used in
+			// statement traces to give more information in statement diagnostic bundles.
+			autoRetryReason error
+
+			// atomicAutoRetryCounter keeps track of the which iteration of a transaction
+			// auto-retry we're currently in. It's 0 whenever the transaction state is not
+			// stateOpen.
+			atomicAutoRetryCounter int32
+		}
```

Comment:
This seems to be fighting the model. While I think this could work, it risks muddling the concepts of the connExecutor to the point that I'm unsure that it'll be easy to walk back. The connExecutor has a field `state          txnState` which already contains synchronization, and, well, a whole framework for how to update it. The `extraTxnState`, as I understand it, is stuff that's related to the txn but not for external reading or synchronization. 

I think the right thing to is to look at the FSM to be setting these fields in the `txnState`. A sketch is:

```go
+++ b/pkg/sql/conn_fsm.go
@@ -530,16 +530,14 @@ func prepareTxnForRetry(args fsm.Args) error {
 }
 
 func prepareTxnForRetryWithRewind(args fsm.Args) error {
+       pl := args.Payload.(eventRetriableErrPayload)
        ts := args.Extended.(*txnState)
        ts.mu.Lock()
        ts.mu.txn.PrepareForRetry(ts.Ctx)
+       ts.mu.autoRetryReason = pl.err
        ts.mu.Unlock()
        // The caller will call rewCap.rewindAndUnlock().
-       ts.setAdvanceInfo(
-               rewind,
-               args.Payload.(eventRetriableErrPayload).rewCap,
-               txnEvent{eventType: txnRestart},
-       )
+       ts.setAdvanceInfo(rewind, pl.rewCap, txnEvent{eventType: txnRestart})
        return nil
 }
 
diff --git a/pkg/sql/txn_state.go b/pkg/sql/txn_state.go
index 253b756795..92e85b83a1 100644
--- a/pkg/sql/txn_state.go
+++ b/pkg/sql/txn_state.go
@@ -60,6 +60,8 @@ type txnState struct {
                // stmtCount keeps track of the number of statements that the transaction
                // has executed.
                stmtCount int
+
+               autoRetryReason error
        }

```


I vote we also update the rule of thumb on the struct to indicate that if we need to export the information via `serialize` then we need it to be in `txnState`.

Reply:
Done.